chain/client/src/client.rs

```diff
@@ -320,6 +346,73 @@ impl ClientActor {
         if status == BlockStatus::Next || status == BlockStatus::Reorg {
             self.tx_pool.reconcile_block(&block);
         }
+
+        self.check_announce_account(&block);
+    }
+
+    fn check_announce_account(&mut self, block: &Block) {
+        // Announce AccountId if client is becoming a validator soon.
+        // First check that we currently have an AccountId
+
+        if self.block_producer.is_none() {
+            // There is no account id associated with this client
+            return;
+        }
+
+        let block_producer = self.block_producer.as_ref().unwrap();
+
+        // TODO: Review this formula.
+        let (epoch_hash, _offset) = match self.runtime_adapter.get_epoch_offset(
+            block.hash(),
+            block.header.height + self.config.announce_account_horizon,
+        ) {
+            Ok((epoch_hash, offset)) => (epoch_hash, offset),
+            // Don't announce if the block is unknown to us
+            Err(_) => return,
+        };
+
+        if let Ok(epoch_block) = self.chain.get_block(&epoch_hash) {
+            let epoch_height = epoch_block.header.height;
+
+            if self.last_val_announce_epoch.is_some()
+                && self.last_val_announce_epoch.unwrap() >= epoch_height
```

Comment:
it's better to not use `unwrap()` explcitly instead `if let Some(last_announce) = self.last_val_announce_epoch { if last_announce >= epoch_height { return } }`